### PR TITLE
Add support for keyword "explode" in an operation's parameter

### PR DESCRIPTION
**Pull request by nikolaas**
Proposed fix for issue #68.

I also implemented reading of a parameter's value according to a serialization method as it defined is the specification [(Describing parameters](https://swagger.io/docs/specification/describing-parameters/) and [Parameter serialization](https://swagger.io/docs/specification/serialization/))

### CODE REVIEW

File: openapi3/parameter.go [hunk 1]
```diff
@@ -130,6 +141,39 @@ func (parameter *Parameter) UnmarshalJSON(data []byte) error {
 	return jsoninfo.UnmarshalStrictStruct(data, parameter)
 }
 
+type SerializationMethod struct {
+	Style   string
+	Explode bool
+}
+
+var supportedSerializationMethods = map[string][]SerializationMethod{
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3/parameter.go]: Done.

File: openapi3/parameter.go [hunk 2]
```diff
@@ -144,6 +188,21 @@ func (parameter *Parameter) Validate(c context.Context) error {
 	default:
 		return fmt.Errorf("Parameter can't have 'in' value '%s'", parameter.In)
 	}
+
+	// Validate a parameter's serialization method.
+	sm := parameter.SerializationMethod()
+	supported := false
+	for _, v := range supportedSerializationMethods[in] {
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3/parameter.go]: Done.

File: openapi3/parameter.go [hunk 3]
```diff
@@ -160,3 +219,33 @@ func (parameter *Parameter) Validate(c context.Context) error {
...
+		if style == "" {
+			style = SerializationForm
+		}
+		explode := true
+		if parameter.Explode != nil {
+			explode = *parameter.Explode
+		}
+		return SerializationMethod{Style: style, Explode: explode}
+	default:
+		panic(fmt.Sprintf("unexpected parameter's 'in': %q", parameter.In))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3/parameter.go]: Done.

File: openapi3filter/param_decoder.go [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+	case openapi3.ParameterInPath:
+		decoder = &pathParamDecoder{input: input}
+	case openapi3.ParameterInQuery:
+		decoder = &queryParamDecoder{input: input}
+	case openapi3.ParameterInHeader:
+		decoder = &headerParamDecoder{input: input}
+	case openapi3.ParameterInCookie:
+		decoder = &cookieParamDecoder{input: input}
+	default:
+		panic(fmt.Sprintf("unsupported parameter's 'in': %s", param.In))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+	var prefix string
+	switch sm.Style {
+	case "simple":
+		// A prefix is empty for style "simple".
+	case "label":
+		prefix = "."
+	case "matrix":
+		prefix = ";" + param.Name + "="
+	default:
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+type headerParamDecoder struct {
+	input *RequestValidationInput
+}
+
+// DecodePrimitive decodes a raw value of header parameter to a value of a primitive type
+// according to rules of the OpenAPI 3 specification.
+func (d *headerParamDecoder) DecodePrimitive(param *openapi3.Parameter) (interface{}, error) {
+	sm := param.SerializationMethod()
+	if sm.Style != "simple" {
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+	}
+
+	values := d.input.Request.Header[http.CanonicalHeaderKey(param.Name)]
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+		return nil, nil
+	}
+	val, err := parsePrimitive(values[0], param.Schema)
+	if err != nil {
+		return nil, &RequestError{Input: d.input, Parameter: param, Reason: errMsgInvalidValue, Err: err}
+	}
+	return val, nil
+}
+
+// DecodeArray decodes a raw value of header parameter to an array according to rules of the OpenAPI 3 specification.
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+// DecodeArray decodes a raw value of cookie parameter to an array according to rules of the OpenAPI 3 specification.
+func (d *cookieParamDecoder) DecodeArray(param *openapi3.Parameter) ([]interface{}, error) {
+	sm := param.SerializationMethod()
+	if sm.Style != "form" || sm.Explode {
+		panic(invalidSerializationMsg(param))
+	}
+
+	cookie, err := d.input.Request.Cookie(param.Name)
+	if err == http.ErrNoCookie {
+		// A HTTP request does not contain a corresponding cookie.
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+	return val, nil
+}
+
+// propsFromString returns a properties map that is created by splitting a source string by propDelim and valueDelim.
+// The source string must have a valid format: pairs <propName><valueDelim><propValue> separated by <propDelim>.
+// The function returns an error when the source string has an invalid format.
+func propsFromString(src, propDelim, valueDelim string) (map[string]string, error) {
+	props := make(map[string]string)
+	if propDelim == valueDelim {
+		pairs := strings.Split(src, propDelim)
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 11]
```diff
@@ -0,0 +1,606 @@
...
+		prefix = "."
+		delim = "."
+	case sm.Style == "matrix" && sm.Explode == false:
+		prefix = ";" + param.Name + "="
+		delim = ","
+	case sm.Style == "matrix" && sm.Explode == true:
+		prefix = ";" + param.Name + "="
+		delim = ";" + param.Name + "="
+	default:
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 12]
```diff
@@ -0,0 +1,606 @@
...
+type queryParamDecoder struct {
+	input *RequestValidationInput
+}
+
+// DecodePrimitive decodes a raw value of query parameter to a value of a primitive type
+// according to rules of the OpenAPI 3 specification.
+func (d *queryParamDecoder) DecodePrimitive(param *openapi3.Parameter) (interface{}, error) {
+	sm := param.SerializationMethod()
+	if sm.Style != "form" {
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 13]
```diff
@@ -0,0 +1,606 @@
...
+				props[groups[0][1]] = values[0]
+			}
+			if len(props) == 0 {
+				// A HTTP request does not contain query parameters encoded by rules of style "deepObject".
+				return nil, nil
+			}
+			return props, nil
+		}
+	default:
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 14]
```diff
@@ -0,0 +1,606 @@
...
+		return nil, &RequestError{Input: d.input, Parameter: param, Reason: errMsgInvalidValue, Err: err}
+	}
+	return val, nil
+}
+
+// DecodeArray decodes a raw value of query parameter to an array according to rules of the OpenAPI 3 specification.
+func (d *queryParamDecoder) DecodeArray(param *openapi3.Parameter) ([]interface{}, error) {
+	sm := param.SerializationMethod()
+	if sm.Style == "deepObject" {
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

File: openapi3filter/param_decoder.go [hunk 15]
```diff
@@ -0,0 +1,606 @@
...
+	case sm.Style == "matrix" && sm.Explode == false:
+		prefix = ";" + param.Name + "="
+		propsDelim = ","
+		valueDelim = ","
+	case sm.Style == "matrix" && sm.Explode == true:
+		prefix = ";"
+		propsDelim = ";"
+		valueDelim = "="
+	default:
+		panic(invalidSerializationMsg(param))
```
[the remark this answers is not in the document]
Comment (nikolaas) [reply openapi3filter/param_decoder.go]: Done.

**Comment by nikolaas** [comment]
@fenollp I fixed all your comments.

File: openapi3filter/param_decoder_test.go [hunk 16]
```diff
@@ -0,0 +1,1041 @@
...
+			},
+		},
+	}
+
+	for _, tg := range testGroups {
+		t.Run(tg.name, func(t *testing.T) {
+			for _, tc := range tg.testCases {
+				t.Run(tc.name, func(t *testing.T) {
+					req, err := http.NewRequest(http.MethodGet, "http://test.org/test"+tc.path, nil)
+					if err != nil {
```
Comment (fenollp) [inline openapi3filter/param_decoder_test.go]: From here and until the end of the file, could you use `"github.com/stretchr/testify/require"` instead?

Comment (fenollp) [reply openapi3filter/param_decoder_test.go]: See here https://github.com/getkin/kin-openapi/blob/e6aa6352832d5c2b0bc28bc5e1065db6cc18d221/openapi3/schema_test.go#L52 how it makes the assumptions a bit more readable.

Comment (nikolaas) [reply openapi3filter/param_decoder_test.go]: Done.

**Comment by nikolaas** [comment]
@fenollp I done it.

**Comment by fenollp** [comment]
Amazing! Thanks

**Pull request merged**